Article about 'Thinking in Iterators' (#21)
corrode/corrode.github.io: content/blog/list-comprehensions/index.md

comprehensions](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions).

```python
# A list of bands
bands = [
    "Metallica",
    "Iron Maiden",
    "AC/DC",
    "Judas Priest",
    "Megadeth"
]

# A list comprehension to filter for bands that start with "M"
m_bands = [band for band in bands if band.startswith("M")]

# A list comprehension to uppercase the bands
uppercased = [band.upper() for band in m_bands]

# This gives us ["METALLICA", "MEGADETH"] 
```

List comprehensions are a concise way to create lists from other lists,
but they work with any iterable, like dicts!

```python
books = {
    "The Lord of the Rings": "J. R. R. Tolkien",
    "The Hobbit": "J. R. R. Tolkien",
    "Harry Potter": "J. K. Rowling"
}
tolkien_books = [
    book for book, author in books.items()
    if "Tolkien" in author
]
# tolkien_books = ["The Lord of the Rings", "The Hobbit"]
```

As a Pythonista, list comprehensions become second nature to me.
Their elegance is hard to beat. For a long time, I wished Rust had something
similar. It was one of the few things I *profoundly* missed from Python &mdash;
until I learned more about the philosophy behind Rust's iterator patterns.

> This is what this post is about.

Reviewer: "This" is a bit unclear. This revelation, this journey, this philosophy. Something like that.

Good intro otherwise.

Author: Thanks. Done.

> This journey is what this post is about.

Reviewer: Still sounds weird :-/

Author: Maybe I just remove it?

Author: removed

operations are [`map`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.map), which applies a function to each item, and [`filter`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter), which
selectively includes items that meet a certain condition.

Here's the equivalent Rust code for the first Python example above:

```rust
let bands = vec![
    "Metallica",
    "Iron Maiden",
    "AC/DC",
    "Judas Priest",
    "Megadeth",
];

let uppercased: Vec<_> = bands.iter()
                    .filter(|band| band.starts_with("M"))
                    .map(|band| band.to_uppercase())
                    .collect();

// uppercased = vec!["METALLICA", "MEGADETH"]
```

Sure, it's a tad more verbose than Python's list comprehensions, but oh, the
versatility!

For example, debugging an iterator chain is as simple as inserting an `inspect`
wherever you want to peek at the elements:

```rust
let uppercased: Vec<_> = bands
    .iter()
    .filter(|band| band.starts_with("M"))
    .inspect(|band| println!("Found band that starts with M: {band}"))
    .map(|band| band.to_uppercase())
    .collect();
```

> This would be quite tricky to do with list comprehensions in Python.

Reviewer: Many "this" in the post ;-)

```suggestion
Achieving the same effect with list comprehensions in Python is quite tricky.
```

Author: OK. I'll go through the article again and look for replacements.

I often find myself chaining iterator operations in Rust. And honestly? It's
pretty intuitive. Over time, I've even grown fond of its explicitness. (Or, you know,
maybe it's just some form of [Stockholm syndrome](https://en.wikipedia.org/wiki/Stockholm_syndrome).)

What I like the most is the flexibility of this pattern. Let me show you!

## Collecting into different types

In Python, you can collect into different types like [`set`](https://docs.python.org/3/tutorial/datastructures.html#sets):

```python
tolkien_books = {
    book for book, author in books.items()
    if "Tolkien" in author
}
# It's a set!
# tolkien_books = {"The Lord of the Rings", "The Hobbit"}

> ```

Reviewer: `set` is a bit weird, especially since it uses curly braces. Maybe a dict would be better? 🤷

Author: Wanted to show one of the lesser used datastructures for a change. The idea was to show that list comprehension are quite flexible.

    .map(|(book, _)| book)
    .collect();
```

What if we want to count the number of bands that start with the same letter?

```rust
let first_letters: HashMap<char, usize> = bands
    .iter()
    // Filter out bands that don't have a first letter.
    // `filter_map` is like `map` but it filters out `None` values.
    // If the band is empty, `chars().next()` will return `None`.
    .filter_map(|name| name.chars().next())
    // Start counting the occurrences of each letter
    .fold(HashMap::new(), |mut acc, c| {
        // Use the entry API to insert a new key or increment the value
        // https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html
        // This gets the entry for the character `c` and inserts 0 if it
        // doesn't exist
        *acc.entry(c).or_insert(0) += 1;
        acc
    });

// Printing the result
for (letter, count) in &first_letters {
    println!("{}: {}", letter, count);
}
```

This gives us:

```text
M: 2
I: 1
A: 1
J: 1
```

In Python, you would probably stop using list comprehensions at this point and
use the builtin `Counter` for this:

```python
from collections import Counter

> first_letters = Counter([band[0] for band in bands])
> # Note that we didn't handle the case where a band 
> # doesn't have a first letter

Reviewer: ```suggestion
first_letters = Counter([band[0] for band in bands if len(band)])
```

ez

Author: Thanks
